keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,294 @@
+    * After deprecation period, scheduler won't schedule any pods that use volumes to a node with missing `CSINode` instance. It is expected that it happens only during node registration when `Node` exists and `CSINode` doesn't and it self-heals quickly.
+
+* `CSINode.status.allocatable` is array of limits. Following combinations are possible:
+
+|  `Volumes` | Description  |
+|  --------- | ------------ |
+|  0         | plugin / CSI driver exists and has zero limit, i.e. can attach no volumes |
+|  X>0       | plugin / CSI driver exists and can attach X volumes (where X > 0) |
+|  X<0       | negative values are blocked by validation
+| Driver is missing in `CSINode.status.allocatable`   | there is no limit of volumes on the node* |
+
+*) This way we are not able to distinguish between a volume plugin / CSI driver not installed on a node or it has been installed and it has no limits.
+  * Predicates modified in this KEP assume that storage provided by **in-tree** volume plugin is available all nodes in the cluster. Other predicate(s) will evaluate `PV.spec.nodeAffinity` and filter out nodes that don't have access to the storage.
+  * For CSI drivers, availability of a CSI driver on a node can be checked in `CSINode.spec`. Its handling is out of scope of this KEP, see non-goals.
+
+CSINode example:
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: CSINode
+metadata:
+  name: ip-172-18-4-112.ec2.internal
+spec:
+status:
+  allocatable:
+    # AWS node can attach max. 40 volumes, 1 is reserved for the system
+    - name: ebs.csi.aws.com
+      volumes: 39
+```
+
+* Existing scheduler predicates `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount` are already deprecated.
+  * If any of them is enabled together with `MaxCSIVolumeCountPred`, the deprecated predicate will do nothing (`MaxCSIVolumeCountPred` does the job of counting both in-tree and CSI volumes).
+  * The deprecated predicates will do any useful work only when `MaxCSIVolumeCountPred` predicate is disabled or `CSINode` object does not have limits for a particular driver.
+  * This way, we save CPU by running only one volume limit predicate during deprecation period.
+
+### New API
+
+CSINode gets `Status` struct with `Allocatable`, holding limit of volumes for each volume plugin and CSI driver that can be scheduled to the node.
+
+```go
+type CSINode struct {
+	...
+
+	// spec is the status of CSINode
+	Status CSINodeStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// CSINodeStatus holds information about the status of all CSI drivers installed on a node
+type CSINodeStatus struct {
+	// allocatable is a list of volume limits for each volume plugin and CSI driver on the node.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Allocatable []VolumeLimits `json:"allocatable" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=allocatable"`
```

Comment:
Should this be a CSINodeDriverStatus to mirror spec more closely?

Is it odd that a status for a driver may exist before the spec? Although in the past designs iterations for migration we had confusing behavior where driver spec would be partially filled in before driver was installed

Reply:
I think we want to avoid having a partially filled spec, both are anti-patterns. That leaves this style as the only real option IMO, but that means if we ever add a "real status field" we have to be ok with the pretty weird schema of:
```yaml
apiVersion: storage.k8s.io/v1beta1
kind: CSINode
metadata:
  name: ip-172-18-4-112.ec2.internal
spec:
status:
  drivers:
    - name: ebs.csi.aws.com
    fieldA: valueA
  allocatable:
    # AWS node can attach max. 40 volumes, 1 is reserved for the system
    - name: ebs.csi.aws.com
      volumes: 39
```
where drivers are duplicated between the allocatable and drivers fields.

What is the API guidance on having a `Status` without a corresponding `Spec`. Is that even weirder? @liggitt

Reply:
I am not sure why is having a `Status` field without having a corresponding `Spec` field weird. `Node` Object is a good example and `CSINode` object in-fact mirrors properties of a node.

Reply:
We may end up with two arrays with the same keys and that would be indeed weird. We could do:

```yaml
status:
  drivers:
    - name: ebs.csi.aws.com
      fieldXYZ: x
      allocatable:
          volumes: 39
          size: "64 TB"
```

But that would complicate sharing of limits between drivers, if we ever implement it.

Reply:
On a meeting we decided to move `allocatable` into `CSINode.spec.drivers[xyz]`, because it does not change in time.